src/utils/ERC20CosmosCoinAmountConversion.sol

```diff
@@ -0,0 +1,137 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.25;
+
+// https://docs.openzeppelin.com/contracts/4.x/api/token/erc20#IERC20Metadata
+import { IERC20Metadata } from "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+
+library ERC20CosmosCoinAmountConversion {
+    // Using Constants for decimals
+    uint8 constant DEFAULT_ERC20_DECIMALS = 18;
+    // https://docs.cosmos.network/v0.50/build/architecture/adr-024-coin-metadata
+    // TODO Do we want to support flexibility for cosmos coin decimals?
+    uint8 constant DEFAULT_COSMOS_DECIMALS = 6;
+
+    // Note that ERC20 standard use 18 decimals by default. Custom ERC20 implementation may decide to change this.
+    /**
+     * @notice Gets the decimals of a token if it extends the IERC20 standard using IERC20Metadata
+     * @param tokenAddress The address of the token contract
+     * @return decimals The number of decimals of the token
+     */
+    function _getERC20TokenDecimals(address tokenAddress) internal view returns (uint8) {
+        // Input validation
+        // TODO Add custom error?
+        require(tokenAddress != address(0), "Address cannot be the zero address");
+        // TODO Reason about this check - Remember it as a good practice but what real protections give us?
+        // TODO Add custom error?
+        require(tokenAddress != address(this), "Address cannot be the contract itself");
+
+        // If the tokens extends the IERC20 it should be using IERC20Metadata which supports the decimals() call
+        // Why this? -->  https://detectors.auditbase.com/decimals-erc20-standard-solidity
+        try IERC20Metadata(tokenAddress).decimals() returns (uint8 decimals) {
+            return decimals;
+        } catch {
+            // If the decimals function is not implemented, return the default value
+            return DEFAULT_ERC20_DECIMALS;
+        }
+    }
+
+    // Convert ERC20 tokens decimals to Cosmos decimals units
+    // NOTE that the refund to the user of the remainder of the conversion should happen during the transfer.
+    /**
+     * @notice Convert the uint256 ERC20 token amount into cosmos coin amount uint64
+     * @notice This functions allows us to consider the loss of precision
+     * @param tokenAddress The address of the token contract
+     * @param amount The amount to be converted
+     * @return convertedAmount The amount converted to uint64 supported by cosmos coins
+     * @return remainder The remainder of the conversion
+     */
+    function _convertERC20AmountToCosmosCoin(
+        address tokenAddress,
+        uint256 amount
+    )
+        internal
+        view
+        returns (uint64 convertedAmount, uint256 remainder)
```

Comment:
In general, I don't like using named return variables. I suggest we remove them for good or add a linter rule
```suggestion
        returns (uint64, uint256)
```

Reply:
Out of curiosity, why don't you like them? In scenarios where it is not right away obvious what the values are going to represent (like here), I kind of like having them named personally. But I am OK with having a rule to not use named return variables also :) Just curious about the reasoning.

Reply:
Yeah, let's decide on this. I find naming them useful for clarify and even because sometimes you can directly change their value if the return value needs to be updated multiple times without creating an additional one.